conjure-go-client/httpclient/body_handler.go

```diff
@@ -60,10 +60,7 @@ func (b *bodyMiddleware) setRequestBody(req *http.Request) error {
 	// use the provided input directly as the request body.
 	if bodyReadCloser, ok := b.requestInput.(io.ReadCloser); ok && b.requestEncoder == nil {
 		req.Body = bodyReadCloser
-		// Use the same heuristic as http.NewRequest to generate the "GetBody" function.
-		if newReq, err := http.NewRequest("", "", bodyReadCloser); err == nil {
-			req.GetBody = newReq.GetBody
-		}
+		req.GetBody = nil
```

Comment:
In the cases where the ReadCloser is something resettable (e.g. bytes.Buffer), we want GetBody to be set so the client can correctly retry the request.

From the [godoc](https://golang.org/pkg/net/http/#NewRequestWithContext):
> If body is of type *bytes.Buffer, *bytes.Reader, or *strings.Reader, the returned request's ContentLength is set to its exact value (instead of -1), GetBody is populated (so 307 and 308 redirects can replay the body), and Body is set to NoBody if the ContentLength is 0.

Reply:
Ah, got it, that is true. In that case do we even need this line or will it just default to nil?

Anything we can do to make it retryable in the case that it really is stored in memory (without forcing in-memory caching for readers that are read from network/disk)? Theoretically we could loosen the argument from ReadCloser to Reader, leave or duplicate the `NewRequest` type switch, and add a NopCloser if it's not a closer.